### ISSUE TITLE
Skipping rate limiter in test does not work

### ISSUE THREAD

**Issue by blimmer** [ParabolInc/parabol#6381]
## Issue - Bug

- **How serious is the bug?** P3
- **OS:** all
- **Browser:** N/A

The current behavior to disable rate limiting in test does not work as expected. This is causing failures in https://github.com/ParabolInc/parabol/pull/6183.

Originally, I had implemented the RateLimiter as configurable via the `.env` file (see https://github.com/ParabolInc/parabol/pull/6063). However, I received feedback that we should instead use the `NODE_ENV` environment variable to disable to rate limiter instead (see superseding PR https://github.com/ParabolInc/parabol/pull/6146).

The `NODE_ENV` approach causes problems in the current testing behavior (where we use the development server) and the proposed changes in https://github.com/ParabolInc/parabol/pull/6379, which uses the production build for testing.

In both dev and CI, `process.env.NODE_ENV` is built into the code at compile time by Webpack. So, the environment variable is not checked at runtime.

### Problem in Dev

In dev, the environment variable approach causes problems because the `NODE_ENV` is compiled into the code by Webpack. This makes it difficult for users to easily switch between `development` and `test` `NODE_ENV` values (they must recompile). Additionally, the `NODE_ENV` variable does not appear to be passed to the webpack build in dev (e.g., `NODE_ENV=test yarn build` does not replace `process.env.NODE_ENV` with `test` during the build process).

### Problem in CI

In CI, we have a similar issue today and will also have the same issue once https://github.com/ParabolInc/parabol/pull/6379 is complete. The goal of https://github.com/ParabolInc/parabol/pull/6379 is good - using the `production` build for testing. However, the approach of using the `NODE_ENV` environment variable breaks down that strategy. We'd have to choose **either** testing against a more production-like environment __or__ disabling rate limiting. Many checks through that code check if the environment is production, so tests will be less prod-like following the current disable approach from https://github.com/ParabolInc/parabol/pull/6146.

### StubRateLimiter

We could change the rate-limiting disabling logic back to the initially proposed solution from https://github.com/ParabolInc/parabol/pull/6063. This will allow us to selectively disable rate limiting via the `.env` file in CI and Dev.

### Dyanmically Reference `NODE_ENV` vs. Building via Webpack

Alternatively, we could look into dynamically reading `NODE_ENV` from the environment vs. building the variable into the webpack code. However, this would make tests less production-like and would take longer to accomplish.

### Ignore Localhost-y IP Addresses

At least running on my local machine, I see my IP address as `0000:0000:0000:0000:0000:ffff:7f00:0001` (the IPv6 equivalent of `127.0.0.1`). Assuming this is the same in CI, we could ignore that IP address in the rate limiter.

### Future Use-Case

In the future, we'd like to run integration tests as part of the GitLab CI deployment pipeline. I don't think any of these three ideas above will support that workflow, so we might need to brainstorm the future of our rate-limiting to work around this issue in the longer term. 

As we migrate over to a horizontally-scalable application, we'll need to reconsider the current in-memory rate limiting behavior, anyway, as it is "per instance," not global. It might make sense to treat this as two issues - one short-term to get integration tests working and one longer-term for the future of rate limiting in the new infrastructure environment.

**Comment by blimmer** [comment]
@adaniels-parabol and I were also thinking that we'd like to run this test suite as part of the GitLab deployment pipelines. Maybe we need some other method when running in production mode to bypass rate limiting?

**Comment by mattkrick** [comment]
thanks for bringing this up!
I think testing against a prod build seems like a reasonable goal.
webpack does magic on `process.env.NODE_ENV`, and individual packages treat that as magic too (eg react, relay). perhaps we invent a new env var or reuse `CIRCLE_CI`?

Some additional goals:
- only build once (don't bulid once for testing, once for deployment)
- make this robust enough to support other singletons or state-bearing objects
- handle other CI patterns (where IP address may not be localhost)

given those constraints, adding or reusing a different env var seems reasonable (e.g. process.env.CI or process.env.CIRCLE_CI)

From that, whether we use a stub within getRateLimiter or we add conditional logic inside the `log` method of the existing rate limiter both seems OK. I generally dislike stubs, but it could be nice because it means the logic only gets checked once, which feels right. 

basically this is like your original PR but instead of having an env var for just the stub, it is an CI-wide env var that we can reuse in the future

```ts
const StubbedLimiter = {
  log: () => ({lastMinute: 1, lastHour: 1})
}
const getRateLimiter = () => {
  if (!rateLimiter) {
    rateLimiter = process.env.CI ? StubbedLimiter: new InMemoryRateLimiter()
  }
  return rateLimiter
}
```

**Comment by blimmer** [comment]
Sure, for CI we can probably use `process.env.CI` since that's pretty consistent across all providers. The dev workflow will be a bit weird (setting `CI=true` locally to disable rate-limiting), but I can call that out in the README.

This won't solve the future use-case of wanting to test live production instances before swapping ingress over, but we can break out a separate story for that as well.
